File: src/bindings/swig/lua/tests/test_kdb.lua
require("kdb")

TEST_NS = "user:/tests/swig_lua"

-- kdbconfig.h
assert(type(kdb.DB_SYSTEM) == "string")
assert(type(kdb.DB_USER)   == "string")
assert(type(kdb.DB_HOME)   == "string")
assert(type(kdb.DEBUG)     == "number")

-- kdb.h
assert(type(kdb.VERSION) == "string")
assert(type(kdb.VERSION_MAJOR) == "number")
assert(type(kdb.VERSION_MINOR) == "number")
assert(type(kdb.VERSION_PATCH) == "number")
assert(kdb.KS_END == nil)

-- ctor
assert(swig_type(kdb.KDB())      == "kdb::KDB *")
local error = kdb.Key(nil)
kdb.KDB(error)
assert(swig_type(kdb.KDB(error)) == "kdb::KDB *")

-- get
do
	local db = kdb.KDB()
	local ks = kdb.KeySet(100)
	db:get(ks, "system:/elektra")

	if os.getenv("CHECK_VERSION") == nil then
		local key = ks["system:/elektra/version/constants/KDB_VERSION"]
		assert(key.value == kdb.VERSION)
	end
end

-- set
do
	local db = kdb.KDB()
	local ks = kdb.KeySet(100)
	db:get(ks, TEST_NS)

	local key = ks[TEST_NS .. "/mykey"]
	if not key then
		key = kdb.Key(TEST_NS .. "/mykey")
		ks:append(key)
	end
	key.string = "new_value"

	db:set(ks, TEST_NS)
end

do
	local db = kdb.KDB()
	local ks = kdb.KeySet(100)
	db:get(ks, TEST_NS)
	assert(ks[TEST_NS .. "/mykey"].value == "new_value")
end

-- cleanup
do
	local db = kdb.KDB()
	local ks = kdb.KeySet(100)
	db:get(ks, TEST_NS)
	ks:cut(kdb.Key(TEST_NS))
	db:set(ks, TEST_NS)
end

File: src/bindings/swig/python/tests_kdb/test_kdb.py
import kdb, unittest

TEST_NS = "user:/tests/swig_py3"

class Constants(unittest.TestCase):
	def setUp(self):
		pass

	def test_kdbconfig_h(self):
		self.assertIsInstance(kdb.DB_SYSTEM, str)
		self.assertIsInstance(kdb.DB_USER,   str)
		self.assertIsInstance(kdb.DB_HOME,   str)
		self.assertIsInstance(kdb.DEBUG,     int)

	def test_kdb_h(self):
		self.assertIsInstance(kdb.VERSION,       str)
		self.assertIsInstance(kdb.VERSION_MAJOR, int)
		self.assertIsInstance(kdb.VERSION_MINOR, int)
		self.assertIsInstance(kdb.VERSION_PATCH, int)
		self.assertIsNone(kdb.KS_END)

class KDB(unittest.TestCase):
	def test_ctor(self):
		self.assertIsInstance(kdb.KDB(), kdb.KDB)
		error = kdb.Key()
		self.assertIsInstance(kdb.KDB(error), kdb.KDB)

	def test_get(self):
		with kdb.KDB() as db:
			ks = kdb.KeySet()
			db.get(ks, "system:/elektra")

			import os
			if os.getenv("CHECK_VERSION") is None:
				key = ks["system:/elektra/version/constants/KDB_VERSION"]
				self.assertEqual(key.value, kdb.VERSION)

	def test_set(self):
		with kdb.KDB() as db:
			ks = kdb.KeySet(100)
			db.get(ks, TEST_NS)

			try:
				key = ks[TEST_NS + "/mykey"]
			except KeyError:
				key = kdb.Key(TEST_NS + "/mykey")
				ks.append(key)
			key.value = "new_value"

			db.set(ks, TEST_NS)

		with kdb.KDB() as db:
			ks = kdb.KeySet(100)
			db.get(ks, TEST_NS)
			self.assertEqual(ks[TEST_NS + "/mykey"].value, "new_value")

	def test_diff(self):
		with kdb.KDB() as db:
			ks = kdb.KeySet(100)
			db.get(ks, TEST_NS)

			key = kdb.Key(TEST_NS + "/diffkey")
			key.value = "new_value"
			ks.append(key)

			diff: kdb.ElektraDiff = db.calculateChanges(ks, TEST_NS)

			self.assertFalse(diff.isEmpty())
			self.assertEqual(diff.getAddedKeys()[TEST_NS+"/diffkey"].value, "new_value")

	@classmethod
	def tearDownClass(cls):
		# cleanup
		with kdb.KDB() as db:
			ks = kdb.KeySet(100)
			db.get(ks, TEST_NS)
			ks.cut(kdb.Key(TEST_NS))
			db.set(ks, TEST_NS)

if __name__ == '__main__':
	unittest.main()
